sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/state/AdaptingCircularBufferCounter.java

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.sdk.metrics.internal.state;
+
+/**
+ * A circle-buffer-backed exponential counter.
+ *
+ * <p>The first recorded value becomes the 'baseIndex'. Going backwards leads to start/stop index
+ *
+ * <p>This expand start/End index as it sees values.
+ *
+ * <p>This class is NOT thread-safe. It is expected to be behind a synchronized incrementer.
+ */
+public class AdaptingCircularBufferCounter implements ExponentialCounter {
+  private static final int NULL_INDEX = Integer.MIN_VALUE;
+  private int endIndex = NULL_INDEX;
+  private int startIndex = NULL_INDEX;
+  private int baseIndex = NULL_INDEX;
+  private final AdaptingIntegerArray backing;
+
+  /** Constructs a circular buffer that will hold at most {@code maxSize} buckets. */
+  public AdaptingCircularBufferCounter(int maxSize) {
+    this.backing = new AdaptingIntegerArray(maxSize);
+  }
+
+  /** (Deep)-Copies the values from another exponential counter. */
+  public AdaptingCircularBufferCounter(ExponentialCounter toCopy) {
+    // If toCopy is an AdaptingCircularBuffer, just do a copy of the underlying array
+    // and baseIndex.
+    if (toCopy instanceof AdaptingCircularBufferCounter) {
+      this.backing = new AdaptingIntegerArray(((AdaptingCircularBufferCounter) toCopy).backing);
```

Comment:
`toCopy` indicates we're supposed to copy but this doesn't appear to copy, is it safe?

Reply:
It does copy, it's using the convention of passing the thing to copy into the consstructor, so:  `new AdaptingIntegerArray(otherIntegerArray)` performs a copy.  Is there a different convention I should be using for that?

Reply:
They actually both copy, just via different mechanism.

This one copies using built-in copy-constructors (of backing).   The other copies algorithmically via looping over indexing and setting counts.  This version of the if should be more efficient as it actually does an Arrays.copyOf.

Reply:
I think having a `.copy` method would be cleaner, or otherwise even a comment would be fine too. Currently this code block is not obvious that it is copying due to the significant difference in pattern between the two if / else branches

Reply:
This reads much better, thanks. The issue with the comment is it says "do a copy" without the logic for the copy being there. If there was a comment "Constructor will copy the passed-in array" a bit below it, that could have been fine, but the copy method seems good.